test/integration/etcd/cve_2021_29923_test.go

```diff
@@ -0,0 +1,202 @@
+		},
+		gvr("networking.k8s.io", "v1", "networkpolicies"): {
+			Stub:             `{"metadata": {"name": "np2"}, "spec": {"egress":[{"ports":[{"port":5978,"protocol":"TCP"}],"to":[{"ipBlock":{"cidr":"10.0.012.0/24"}}]}],"ingress":[{"from":[{"ipBlock":{"cidr":"172.017.0.0/16","except":["172.17.001.0/24"]}},{"podSelector":{"matchLabels":{"role":"frontend"}}}],"ports":[{"port":6379,"protocol":"TCP"}]}],"podSelector":{"matchLabels":{"role":"db"}},"policyTypes":["Ingress","Egress"]}}`,
+			ExpectedEtcdPath: "/registry/networkpolicies/" + testNamespace + "/np2",
+		},
+	}
+
+	kindSeen := sets.NewString()
+	pathSeen := map[string][]schema.GroupVersionResource{}
+	etcdSeen := map[schema.GroupVersionResource]empty{}
+	cohabitatingResources := map[string]map[schema.GroupVersionKind]empty{}
+
+	for _, resourceToPersist := range apiServer.Resources {
+		if _, ok := etcdStorageData[resourceToPersist.Mapping.Resource]; !ok {
+			continue
+		}
+		t.Run(resourceToPersist.Mapping.Resource.String(), func(t *testing.T) {
+			mapping := resourceToPersist.Mapping
+			gvk := resourceToPersist.Mapping.GroupVersionKind
+			gvResource := resourceToPersist.Mapping.Resource
+			kind := gvk.Kind
+
+			if kindAllowList.Has(kind) {
+				kindSeen.Insert(kind)
+				t.Skip("allowlisted")
+			}
+
+			etcdSeen[gvResource] = empty{}
+			testData, hasTest := etcdStorageData[gvResource]
+
+			if !hasTest {
+				t.Fatalf("no test data for %s.  Please add a test for your new type to GetEtcdStorageData().", gvResource)
+			}
+
+			if len(testData.ExpectedEtcdPath) == 0 {
+				t.Fatalf("empty test data for %s", gvResource)
+			}
+
+			shouldCreate := len(testData.Stub) != 0 // try to create only if we have a stub
+
+			var (
+				input *metaObject
+				err   error
+			)
+			if shouldCreate {
+				if input, err = jsonToMetaObject([]byte(testData.Stub)); err != nil || input.isEmpty() {
+					t.Fatalf("invalid test data for %s: %v", gvResource, err)
+				}
+				// unset type meta fields - we only set these in the CRD test data and it makes
+				// any CRD test with an expectedGVK override fail the DeepDerivative test
+				input.Kind = ""
+				input.APIVersion = ""
+			}
+
+			all := &[]cleanupData{}
+			defer func() {
+				if !t.Failed() { // do not cleanup if test has already failed since we may need things in the etcd dump
+					if err := client.cleanup(all); err != nil {
+						t.Fatalf("failed to clean up etcd: %#v", err)
+					}
+				}
+			}()
+
+			if err := client.createPrerequisites(apiServer.Mapper, testNamespace, testData.Prerequisites, all); err != nil {
+				t.Fatalf("failed to create prerequisites for %s: %#v", gvResource, err)
+			}
+
+			if shouldCreate { // do not try to create items with no stub
+				if err := client.create(testData.Stub, testNamespace, mapping, all); err != nil {
+					t.Fatalf("failed to create stub for %s: %#v", gvResource, err)
+				}
+			}
+
+			output, err := getFromEtcd(apiServer.KV, testData.ExpectedEtcdPath)
```

Comment:
what exactly do we have to check etcd? Isn't pure creation enough? We are only checking validation, don't we?

Reply:
doesn't this verify that the data is not mutated when persisted?
if I create 10.1.002.3, we should have 10.1.002.3 in etcd and not 10.1.2.3 .... or is it ok to canonicalize it? I think we discussed that and you warned about possible risks of doing that